### ISSUE TITLE
Changes not immediately shown in taskbar & popup's menu item 

### ISSUE THREAD

**Issue by mariospr** [projecthamster/hamster-shell-extension#205]
As mentioned in https://github.com/projecthamster/hamster-shell-extension/issues/189#issuecomment-336856348, when I enter a new task, or stop one in progress, from the popup menu I don't get "immediate feedback": I would expect to see the new/stopped task reflected both in the text showing up in the top panel and in the menu item in the popup menu, but instead I see no change until a minute has passed (at which time the callback kicks-in and updates the text).

STEPS TO REPRODUCE:
 
 1. Create a new task, either from the Popup menu or from Hamster's Overview window

OR, if there's a task already in progress:
 
 1. Click in "Show overview" from the Popup menu
  2. Right click on the task currently in progress then select "Remove"

EXPECTED OUTCOME:

Changes should be immediately reflected in both the extension's popup panel and the taskbar: for newly-created tasks, a new item is shown in the popup and the task's name is shown in the shell's taskbar. For removed items, the task should no longer be visible in the popup panel and, if the removed task was "in progress" the task bar should reset its text to "No activity"

ACTUAL OUTCOME:

No visual change happens in the taskbar nor in the popup panel until a minute has passed since the creation/removal of the aforementioned task

**Comment by elbenfreund** [comment]
Can you confirm that this issue exists with current develop branch?
I setup a fresh debian 9 installation for testing and things seem to work as expected. So I just want to make sure this is not some older issue...

**Comment by mariospr** [comment]
I saw it on my Fedora 26 running from the `develop` branch as it was the day this issue was reported. Perhaps it's some weird interaction with the version of the shell in Fedora 26? Seems unlikely, but it could be... In any case I can confirm it's happening consistently.

**Comment by lgpasquale** [comment]
I experience the same issue on Arch Linux (gnome-shell 3.26.1) with the latest `develop` branch

**Comment by elbenfreund** [comment]
So far I fail to reproduce the issue on a debian vm.
Would you be so kind and let know which hamster version you are running?

**Comment by hedayat** [comment]
This is the error I see when I select to stop an active task or continue a previous one:

```
نوامبر 18 17:26:06 hv gnome-shell[18656]: invalid uninstantiatable type '<invalid>' in cast to 'ClutterText'
نوامبر 18 17:26:06 hv gnome-shell[18656]: clutter_text_get_editable: assertion 'CLUTTER_IS_TEXT (self)' failed
نوامبر 18 17:26:06 hv gnome-shell[18656]: clutter_text_get_text: assertion 'CLUTTER_IS_TEXT (self)' failed
نوامبر 18 17:26:06 hv gnome-shell[18656]: clutter_text_set_text: assertion 'CLUTTER_IS_TEXT (self)' failed
نوامبر 18 17:26:06 hv gnome-shell[18656]: JS ERROR: TypeError: layout is null
                                                TodaysFactsWidget<.populateFactsWidget@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/todaysFactsWidget.js:182:17
                                                wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
                                                TodaysFactsWidget<.refresh@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/todaysFactsWidget.js:193:9
                                                wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
                                                FactsBox<.refresh@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/factsBox.js:86:9
                                                wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
                                                _refresh@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/panelWidget.js:187:3
                                                _proxyInvoker/asyncCallback@resource:///org/gnome/gjs/modules/overrides/Gio.js:90:13

```

**Comment by hedayat** [comment]
Much better now, but I experienced a delay between when a task is started and when it is reflected in the top bar once right now (e.g. more than 10 seconds), using the develop branch.

**Comment by rhertzog** [comment]
Yes, me too. While I had zero delay the other evening. Really strange. Can someone re-open the issue ?

**Comment by rhertzog** [comment]
When I have the problem, it looks like we are not getting any dbus signal when we add a new fact. The menu only gets updated when the 60s timer fires up for regular refresh.

**Comment by rhertzog** [comment]
Maybe this is all related to this part:
```js
                // [FIXME]
                // It seems preferable to have a sync methods call in order to
                // avoid race conditions (previously extensions used async
                // version).
                // There is however a reasonable risk this may actually be a bad
                // idea as it may block the entire shell if the request takes too
                // long. This may be particulary likly if the dbus service uses
                // a remote database as persistent storage.
                // We need to come back to this once we have the low hanging fruits
                // covered.
                let activities = controller.apiProxy.GetActivitiesSync('');
```

Can we have a patch with an asynchronous call here to test?

**Comment by elbenfreund** [comment]
I hope to have a first go at this within the next 2h

**Comment by elbenfreund** [comment]
The ``feature/async-fix`` branch (commit 80afcfc6d71b06b1ae40dc73023d607e880c037a) provides a very quick and dirty approach. As far as I can tell, the only remaining synchronous calls are the edit, open app, open pref buttons. which should not matter unless pressed.

I shall be curious about your results.
Good night :)

**Comment by hedayat** [comment]
I feel these are related to the changes in this branch:

```
ارس 14 13:48:50 hv gnome-shell[27133]: JS ERROR: TypeError: layout is undefined
                                            TodaysFactsWidget<.populateFactsWidget@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/todaysFactsWidget.js:182:17
                                            wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
                                            TodaysFactsWidget<.refresh@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/todaysFactsWidget.js:193:9
                                            wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
                                            FactsBox<.refresh@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/factsBox.js:86:9
                                            wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
                                            _refresh@/home/hedayat/.local/share/gnome-shell/extensions/contact@projecthamster.org/widgets/panelWidget.js:187:3
                                            _proxyInvoker/asyncCallback@resource:///org/gnome/gjs/modules/overrides/Gio.js:90:13
```

Also,  `console.log` is called on line 172, which probably should be `global.log()`

**Comment by elbenfreund** [comment]
Thats weired. Can you reproduce these errors? I get nothing of that sort in my VM.. :/

**Comment by hedayat** [comment]
Not right now. I saw those logs when stopping an activity caused gnome shell crash. But I don't know if they were related. I thought they might happen after locking & unlocking screen, but it doesn't happen now. So, maybe it was really related to the crash.

**Comment by elbenfreund** [comment]
I rebased ``async-fix`` on develop (which now includes PR #240). So you may wanna stick with ``async-fix``.

**Comment by hedayat** [comment]
Well, actually this is what I got crashes on (both #240 & async-fix). So, let's see if I got crashes with async-fix. Then I'll try async-fix too.

**Comment by rhertzog** [comment]
My results are not yet 100% conclusive. It worked a long time but I managed to get a delay at some point but then this was after I got other error messages that I fixed in #242 

Now I'm running with what I submitted + your latest `feature/async-fix` branch and we will see if the problem comes back.

**Comment by elbenfreund** [comment]
So i guess we just wait and see for now, right? I would like to avoid to merge too many changes at once as we really just have trail and error via real life usage as a means o testing.
The reason I am asking is to avoid looking inactive :)

PR #242 should be rather risk free to merge right away of cause. Thanks for this one :+1:

**Comment by rhertzog** [comment]
The extension has been working fine for me throughout the whole day. Yes we should definitely merge #242 as soon as it's ready.

**Comment by elbenfreund** [comment]
``develop`` as well as ``feature/async-fix`` have been updated. Please let us know how things are going for you guys.

**Comment by hedayat** [comment]
Seems to be fixed for me in `feature/async-fix` branch. :)

**Comment by elbenfreund** [comment]
@mariospr Did you have a chance to check this branch? Does it solve your issue?

**Comment by mariospr** [comment]
As far as I can see, everything seems to be working as expected in the `feature/async-fix` branch, at least I can't reproduce any of the issues I reported in https://github.com/projecthamster/hamster-shell-extension/issues/205#issue-265783259, which is great, thanks!

FWIW, I've also checked the `develop` branch and things seem to be working well there too.

**Comment by elbenfreund** [comment]
Wonderfull. Thanks for the time to test! :+1:

**Comment by mariospr** [comment]
My pleasure, and sorry for taking so long!
